aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -7,73 +7,252 @@
 #include <c10/core/TensorOptions.h>
 #include <ATen/core/boxing/KernelFunction.h>
 
-namespace at {
-struct Dimname;
-}
+#include <ATen/core/Dimname.h>
 
 namespace c10 {
 namespace impl {
 
+//
+// utils
+//
+
+// is_specialization_of
+//
+template <template <typename...> class T, typename U>
+struct is_specialization_of: std::false_type {};
+
+template <template <typename...> class T, typename... Us>
+struct is_specialization_of<T, T<Us...>>: std::true_type {};
+
+// is_tuple_of_lvalue_refs
+//
+template<class T, bool = is_specialization_of<std::tuple, T>::value>
```

Comment:
not sure what the `bool = ...` is supposed to do. SFINAE? That only works if you have an `enable_if`.

Reply:
it has the same effect (diverting non-tuples to the other specialization) but I'll recast it in the more traditional form